Theory/Theory.txt

```diff
@@ -0,0 +1,728 @@
+memory is memory space that gets allocated by programmers instructions. Objects are stored here along with Global or
+shared variables. The garbage collector works on the heap memory.
+
+
+13. What is a python module?
+A module is a python code files or scripts that contains functions definitions or statements that can be imported into
+a project. They can be built in modules such as datetime or user-defined.
+
+
+14. What is docstring in Python?
+A docstring is a string that comes as the first statement after the definition of a function, class, method or module.
+It is used to document how that specific segment of code works. They are surrounded by three “”
+
+
+15. What is pickling and unpicking in Python?
+Pickling is the process where Python objects are serialised, converted into byte streams or character streams to be
+stored or sent to other python programs. Unpickling is the inverse, where the byte stream is converted into an object
+
+
+16. What are the tools that help to find bugs or perform static analysis?
+There are a few tools that can be installed using pip, to fix bugs or perform static analysis (check code without
+executing it), popular ones include pylint, pyflakes, mypy and pychecker.
+
+
+17. How are arguments passed in Python by value or by reference? Give an example
+Arguments are not passed by either way in Python, but by assignment or object reference (similar to 'by reference')
+So an OBJECT is referenced, which can be assigned a value inside a function. EG:
+
+string = "I am THE string"
+def example(string):
+    string = "I am assigned to string"
+    print("Inside:", string)
+
+example(string)
+print("Outside:", string)
+
+
+18. What are Dictionary and List comprehensions in Python? Provide examples
+They are a way of shortening syntax for creating Dictionaries or lists from other lists or strings:
+
+List comprehension example:
+new_list = [number for number in range(20) if number % 2 == 0]
+
+Dictionary comprehension example:
+times_hundred_dict = {num: num*100 for num in range(1, 11)}
+
+
+19. What is namespace in Python?
+Namespace contains all the variables or names created in a program that objects are assigned to. The built-in namespace
+contains the names of pythons built-in objects such as int, Len, exception; the global namespace contains names
+defined in the global scope or the main program, including modules. Enclosing or local namespace is namespace local
+to a created variable, they exist only until the function is terminated.
+
+
+20. What is pass in Python?
+It is a null statement or a placeholder for future code. When executed it is not ignored but no operation happens.
+It can hold the place where empty code is not allowed in function and class definitions, loops and if statements.
+
+
+21. What is a unit text in. Python?
+A code level test which verifies a very small section (unit) of code functionality.
+
```

Comment:
0.5/1 - here you need to specify you are testing for "correct" code functionality